### PR TITLE
Jsonify sewage plant

### PR DESCRIPTION
**Pull request by chaisawlajatang**
#### Summary

```SUMMARY: Content "Jsonify sewage treatment plant"```

#### Purpose of change

Sewage plant was old, hardcoded and nonrotatable.

#### Describe the solution

Jsonified the plant and completely remade into rotatable 2x2 building with 3x2 basement.

#### Describe alternatives you've considered

Leave everything as is.

#### Additional context

Sulfuric acid, quicklime and sodium hypochlorite are pretty common chemicals on IRL sewage treatment plants, so I added them.
sewage1
sewage2

### CODE REVIEW

File: data/json/overmap/overmap_special/specials.json [hunk 1]
```diff
@@ -129,30 +129,32 @@
...
+      { "point": [ 2, 1, -1 ], "overmap": "sewage_treatment_2_1_-1_north" }
     ],
+    "locations": [ "land" ],
     "connections": [
-      { "point": [ 0, -1, 0 ] },
-      { "point": [ 0, -1, -1 ], "terrain": "sewer", "connection": "sewer_tunnel", "from": [ 0, 0, -1 ] },
-      { "point": [ 2, 1, -1 ], "terrain": "sewer", "connection": "sewer_tunnel", "from": [ 1, 1, -1 ] },
-      { "point": [ 0, 3, -1 ], "terrain": "sewer", "connection": "sewer_tunnel", "from": [ 0, 2, -1 ] },
-      { "point": [ -2, 1, -1 ], "terrain": "sewer", "connection": "sewer_tunnel", "from": [ -1, 1, -1 ] }
+      { "point": [ 0, -1, 0 ], "terrain": "road" },
```
Comment (Night-Pryanik) [inline data/json/overmap/overmap_special/specials.json]: I can suggest adding an auto-connecting road by something like this:
```suggestion
      { "point": [ 0, -2, 0 ], "terrain": "road", "connection": "local_road", "from": [ 0, -1, 0 ] }
```

**Comment by kevingranade** [comment]
Not sure what happened, locally I built and this had a linker error because of something calling draw_sewer() from mapgen.cpp line 2720, but it isn't a problem in the travis build?

**Comment by Night-Pryanik** [comment]
> Linker error because of something calling draw_sewer() from mapgen.cpp line 2720.

It's `draw_sewer()` in `map.h` and `map.cpp`

**Comment by ZhilkinSerg** [comment]
You need to add savegame migration code.

https://github.com/CleverRaven/Cataclysm-DDA/blob/663942f2c56624499bc5742be0ede575b8f08c85/src/savegame.cpp#L378

https://github.com/CleverRaven/Cataclysm-DDA/blob/27882aeb5d48d56b17dee648292b34fdf108fbd5/data/json/obsolete_terrains.json#L1-L5

**Comment by chaisawlajatang** [comment]
About save migration - I'm not sure how to handle it with single-tile -> multitile location.

**Comment by ZhilkinSerg** [comment]
> 
> 
> About save migration - I'm not sure how to handle it with single-tile -> multitile location.

It was not a single tile previously, but migration is tricky. You can probably check what was done during hospital migration (https://github.com/CleverRaven/Cataclysm-DDA/pull/22543).

**Comment by chaisawlajatang** [comment]
I wonder why the check is errored.

**Comment by ZhilkinSerg** [comment]
> 
> 
> I wonder why the check is errored.

That is unrelated to your changes - just some temporary environmental issue on build serer.

**Comment by ZhilkinSerg** [comment]
Savegame migration is erroneous right now:

- top level contains only one tile:

image

- underground part contains invalid overmap terrains (abandoned drive-throughs): 

image

**Comment by ZhilkinSerg** [comment]
The new mapgen itself looks great and the only thing not working properly is savegame migration. I really hope you can fix it.

**Comment by chaisawlajatang** [comment]
Yeah, I'm trying to figure it out how to transform 1-tile building into a multitile one. I migrated that 1-tile bulding because it could serve as a stub or something, coz it had a downway stairs.

**Comment by ZhilkinSerg** [comment]
You should find that single tile, replace it with new one and put surrounding tiles around.

**Comment by Petethegoat** [comment]
Right now, if I understand the current migration infrastructure correctly, there's no provision for expanding an obsolete special into unrelated adjacent terrains (eg fields, forests, roads, part of an adjacent special).

It's extremely simple to add this capability, and I've got a working version of this already. But I don't know how to handle the potential for intersecting specials- a sewage plant adjacent to a hospital will just eat into it with my current code. I can't think of any nice way to handle this situation, so maybe we just call it quits and accept that migration can't be perfect.

Once I've tested my work a bit more, I'll make a PR to your fork, @chaisawlajatang.

**Comment by Petethegoat** [comment]
Okay, PR created:
https://github.com/chaisawlajatang/Cataclysm-DDA/pull/1

This fully migrates the plant and works great in my testing.
Drawbacks are that it will overwrite adjacent overmap tiles- most of the time this will be fine, and if it's next to another special then one of them is going to have to come out on top, and it might as well be the cool new special. This also makes the added infrastructure very very simple, which I think is neat.

**Comment by Petethegoat** [comment]
Oh, and I had some feedback on the design.
Firstly, it's great- feels totally plausible, instantly recognizable as a sewage treatment plant.

I noticed you didn't add a downspout anywhere. I assume that was intentional, to prevent people easily bypassing the door/console to get underground. Unfortunately, the fences close around the edge of the building let people boost themselves up easily anyway.
I'd just suggest adding a locked shed around the ladder on the rooftop, same as some other buildings have.

I also feel like the zombie spawns are a bit high, especially underground. I'd prefer to see them thinned out and replaced with some sewer gators or something, but that's up to you :smile:

**Comment by chaisawlajatang** [comment]
Thanks for the feedback, I added the downsprout and roof access door and lowered zed spawn a bit. 
Looks like this branch became obsolete, and it turned into a weird shit when I tried to update it.

**Comment by Petethegoat** [comment]
Do you need help fixing the branch? Or maybe @ZhilkinSerg or @kevingranade can utilize maintainer powers to fix things up?

**Comment by chaisawlajatang** [comment]
Yeah, I'm requesting some help.

**Comment by anothersimulacrum** [comment]
I've a PR onto your Branch with a branch that you can reset to to fix this (hopefully) https://github.com/chaisawlajatang/Cataclysm-DDA/pull/2. I've given instructions in the PR form for that branch.

**Comment by ZhilkinSerg** [comment]
Merged as #33339.

**Pull request closed without merging**